src/librustc_query_system/query/config.rs

```diff
@@ -23,6 +23,53 @@ pub trait QueryConfig<CTX> {
     type Value: Clone;
 }
 
+pub(crate) struct QueryVtable<CTX: QueryContext, K, V> {
+    pub anon: bool,
+    pub dep_kind: CTX::DepKind,
+    pub eval_always: bool,
+
+    // Don't use this method to compute query results, instead use the methods on TyCtxt
+    pub compute: fn(CTX, K) -> V,
```

Comment:
I see that this comment is already coming from `fn compute`, but it'd be nice if we could have some wrapper new-type for `K` that isn't accessible from most places, but is accessible from `TyCtxt` to avoid accidental usage.

Reply:
This type is `pub(crate)`, so presumably the fields are also only accessible from within the crate..? (This might be more clear if we made the annotations `pub(crate)`, of course.) In any case that seems reasonable "secure" against mis-use to me.